### ISSUE TITLE
Unexpected error in control flow analysis of enums

### ISSUE THREAD

**Issue by tnikodym** [Microsoft/TypeScript#11522]
**TypeScript Version:**  2.0.3

**Code**

```ts
enum MyEnum {
    A,
    B,
}

declare var option: MyEnum;

switch (option) {
    case MyEnum.A:
        break;
    case MyEnum.B:
        break;

    default:
        // error TS2342: An index expression argument must be of type 'string', 'number', 'symbol', or 'any'.
        console.warn(`Not supported: ${MyEnum[option]}`)
}

```

**Expected behavior:**

Should not throw any error. 

Although the default statement will never be executed, it's a future-proof check for when we add more options to the enum. 

**Actual behavior:** 

Throws the following error: 
*error TS2342: An index expression argument must be of type 'string', 'number', 'symbol', or 'any'.*

**Note:**

VSCode shows inferred type of option in the default clause to be "var option: never".

**Comment by ahejlsberg** [comment]
This is working as intended and is an effect of the control flow analysis introduced in 2.0. The compiler knows that you have exhausted all possible values of the enum type in the preceding switch cases and therefore has narrowed to type of `option` to `never`, effectively indicating that this code will never execute. The error seems entirely reasonable since you're indexing the enum with a value that is known to be out of bounds. However, you can use a type assertion if you really want to override the compiler:

```ts
console.warn(`Not supported: ${MyEnum[<MyEnum>option]}`)
```

**Comment by RyanCavanaugh** [comment]
See also http://stackoverflow.com/questions/39419170/how-do-i-check-that-a-switch-block-is-exhaustive-in-typescript
